### PR TITLE
Improve entity "sync configs"

### PR DESCRIPTION
**Pull request by chriszarate**
## What?

Improve the types and shape of the entity "sync configs" to prepare for future improvements to the SyncProvider.

## Why?

This lays the groundwork for:

1. Improved and type-safe `SyncProvider`
2. Improved CRDT "merge logic" that fit the entity data (e.g., Posts) and utilize [Yjs shared types](https://docs.yjs.dev/getting-started/working-with-shared-types)

## How?

- Removes sync configs for entities that (AFAIK) are not editable within the block editor and therefore cannot currently be synced.
- Centralizes sync-related entity config in a single `syncConfig` property.
- Renames properties for symmetry and clarity.
   - `applyChangesToDoc` => `applyChangesToCRDTDoc`
   - `fromCRDTDoc` => `getChangesFromCRDTDoc`
- Allows the entity sync config to transform a "raw" entity record into sync-ready object data via `getInitialObjectData`. This allows entities to compute and sync derivative properties like `blocks` (derived from `content`).
- Provides a `supports` sub-property for future declaration of sync feature support.

For now, many of the implementations remain simple (e.g., `defaultApplyChangesToCRDTDoc`) but will expand and improve in future PRs.

## Testing Instructions

1. Check out this branch.
1. Enable the real-time collaboration experiment: Gutenberg > Experiments.
1. Open a post for editing in two different browser environments.
1. Use your browser's development tools to remove the post editor lock (see screencast in the next section).

Note: This provider uses requests to `admin-ajax.php` for signaling and the initial connection is delayed.

### Testing Instructions for Keyboard

No UI changes in this pull request.

## Screenshots or screencast 

https://github.com/user-attachments/assets/40d3c34b-ca2a-4e82-aee6-6e644d67e1b3

### CODE REVIEW

File: packages/core-data/src/entities.js [hunk 1]
```diff
@@ -335,13 +305,23 @@ async function loadPostTypeEntities() {
...
-					} );
-				},
-				applyChangesToDoc: ( doc, changes ) => {
-					const document = doc.getMap( 'document' );
+				/**
+				 * Is syncing enabled for this entity?
+				 *
+				 * @type {boolean}
+				 */
+				enabled: true,
```
Comment (youknowriad) [inline packages/core-data/src/entities.js]: Why do we need an extra flag, can we assume that the presence of "syncConfig" is enough?

File: packages/core-data/src/entities.js [hunk 2]
```diff
@@ -366,12 +346,49 @@ async function loadPostTypeEntities() {
...
+				/**
+				 * This initial object data represents the data that will be synced via
+				 * the CRDT document, which may differ from the entity record. There may
+				 * be properties that should not be synced, or properties that are
+				 * derived from the record.
+				 *
+				 * @param {import('@wordpress/sync').ObjectData} record
+				 * @return {import('@wordpress/sync').ObjectData} The initial data
+				 */
+				getInitialObjectData: ( record ) => record,
```
Comment (youknowriad) [inline packages/core-data/src/entities.js]: I don't really understand this one, isn't this a duplicate of syncProperties

File: packages/core-data/src/entities.js [hunk 3]
```diff
@@ -366,12 +346,49 @@ async function loadPostTypeEntities() {
...
+
+				/**
+				 * Get the immutable identifier for an entity record.
+				 *
+				 * @param {import('@wordpress/sync').ObjectData} record
+				 * @return {import('@wordpress/sync').ObjectID} The entity's ID
+				 */
+				getObjectId: ( { id } ) => id,
```
Comment (youknowriad) [inline packages/core-data/src/entities.js]: This is a duplicate of the `key` property of the entity config, so I don't think it's needed.

File: packages/core-data/src/entities.js [hunk 4]
```diff
@@ -366,12 +346,49 @@ async function loadPostTypeEntities() {
...
+
+				/**
+				 * The object type for the entity, used to scope CRDT documents.
+				 *
+				 * @type {import('@wordpress/sync').ObjectType}
+				 */
+				objectType: `postType/${ postType.slug }`,
```
Comment (youknowriad) [inline packages/core-data/src/entities.js]: What is this exactly, can't we just use `kind/name` of the current entity instead?

File: packages/core-data/src/entities.js [hunk 5]
```diff
@@ -366,12 +346,49 @@ async function loadPostTypeEntities() {
...
+
+				/**
+				 * Sync features supported by the entity.
+				 *
+				 * @type {Record< string, boolean >}
+				 */
+				supports: {},
```
Comment (youknowriad) [inline packages/core-data/src/entities.js:367]: Can you clarify this a bit more?

File: packages/core-data/src/entities.js (hunk 1 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js]: No, because some entities within an entity type may not support syncing. For example, posts and pages may support collaborative editing but a custom post type may not.

Here is the PR that implements this flag:
https://github.com/WordPress/gutenberg/pull/72147

Comment (youknowriad) [reply packages/core-data/src/entities.js]: But can't we just omit "syncConfig" entirely for these post types?

File: packages/core-data/src/entities.js (hunk 3 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js]: Not sure I follow. The entity config defines the properties of an entity **type** and is shared across all instances of that entity. Most entity configs do not have a `key` property, including this one.

The "object ID" (as used by the sync provider) is equivalent to the entity record's ID, which is not known until it is loaded into the `core-data` store via `getEntityRecord`. This function allows the sync provider to extract the record ID from the entity record it is given. (This function is also not new, it exists in `trunk` as `getSyncObjectId`.)

We could instead have `core-data` actions and resolvers pass in the "object type" (entity kind/name) and "object ID" (record ID). I was following the existing pattern because I thought there was a desire for the entity config to centralize and control sync behavior, including identifiers. Happy to make the change, just let me know.

(I am borrowing the quoted terms from the initial implementation—we can definitely change them as well!)

File: packages/core-data/src/entities.js (hunk 4 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js]: Yes, see above. This property exists in `trunk` as `syncObjectType` and is just carried forward. We could instead require `kind` / `name` to be passed in each time a sync provider method is called. Let me know and I'll happily update the code.

File: packages/core-data/src/entities.js (hunk 5 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js:367]: This type will eventually become:

```ts
supports?: {
	awareness?: boolean;
	crdtPersistence?: boolean;
	undo?: boolean;
};
```

The sync provider will use it to determine whether to implement a sync feature for a given entity record. Example code from future PR:
https://github.com/Automattic/gutenberg/blob/7c9733c5526d3624bc7c253552191ee1be88aa2b/packages/sync/src/provider.ts#L142-L148

File: packages/core-data/src/entities.js (hunk 2 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js]: You're right, it is partially duplicative! I will update the description to remove the reference to excluded properties. 

However, this function is necessary to compute derived properties for the initial load of the persisted entity record (`SyncProvider#bootstrap`). There is currently only one such example: `blocks` (derived from `content`, which is not synced).

Using this example, blocks are not part of the entity record and are not created when the entity is loaded into the store. Instead, `core-data` provides a [hook](https://github.com/WordPress/gutenberg/blob/v19.3.0/packages/core-data/src/hooks/use-entity-block-editor.js#L147) that parses `blocks` from `content`. The block editor uses this hook and then (eventually, maybe) ships back the property as entity record edits.

However, we need to compute and load `blocks` (once) as part of the bootstrapping process. This ensures that we can correctly reconcile the entity with other peers and / or a persisted CRDT doc.

File: packages/core-data/src/entities.js (hunk 1 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js]: We most certainly can! :)

Comment (chriszarate) [reply packages/core-data/src/entities.js]: Done in https://github.com/WordPress/gutenberg/pull/72145/commits/ed4727d01374b1fa45aac15b3a9677deb42c71c0

File: packages/core-data/src/entities.js (hunk 3 above)
Comment (youknowriad) [reply packages/core-data/src/entities.js]: > I thought there was a desire for the entity config to centralize and control sync behavior

Yes, this is the right approach but I think the original implementation was flawed as well in the sense that this configuration already exists in the "entity's config". `core-data` (without sync) need to know what it's the primary key of each record. It assumes it's `id` by default unless a `key` property is specified on the entity config object. In other words, this function is just a duplicate configuration that we can get rid of. (and just pass the full entity config, or augment the "syncConfig" automatically with the "key" property defined in the entity config)

Comment (youknowriad) [reply packages/core-data/src/entities.js]: (same for kind/name).

File: packages/core-data/src/entities.js (hunk 2 above)
Comment (youknowriad) [reply packages/core-data/src/entities.js]: > However, this function is necessary to compute derived properties for the initial load of the persisted entity record (SyncProvider#bootstrap). There is currently only one such example: blocks (derived from content, which is not synced).

This is not just a problem for "sync", the absence of "blocks" on initial fetch is a problem for the whole core-data abstraction. I basically think that we don't have good support/behavior in core data for what we call "transient properties". I had tried to fixed that in the past and add more awareness for these to the framework (see https://github.com/WordPress/gutenberg/pull/57500)

I think instead of `getInitialObjectData`, we should try to bring the idea of #57500 back to life again (having a get/set config for each transient property) at the entities config level but instead of changing how core-data work with these entirely, take a more iterative approach (only use it for sync for now). WDYT?

File: packages/core-data/src/entities.js (hunk 3 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js]: Got it, I'll rework. Thanks for the guidance!

File: packages/core-data/src/entities.js (hunk 2 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js]: The concept makes sense and I agree it's sorely needed. I can work it into the PR.

> take a more iterative approach (only use it for sync for now).

My interpretation:

1. `transientEdits` properties can be `true` (current implementation) or an object with a `read` function property.
2. The `read` function is called by the `core-data` resolver inside the experimental sync flag to materialize the transient properties on the record before passing it to the sync provider.
3. We will forego a `write` function property for now.
   - ~I wonder if `persist` or some other verb would be more appropriate. A `write` verb implies it is the opposite of `read` and should/would be called on each entity record edit. But transient properties are not "de-materialized" until the record is persisted.~ I think I talked myself out of this. :) `read` is also performed once on entity fetch, so they are in fact opposites.

Comment (youknowriad) [reply packages/core-data/src/entities.js]: Let's keep it contained for now and only call these "read/write" where you'd have previously used `getInitialObjectData`. I think there are other parts of core-data that could benefit from it but we should probably leave that separate as it could be a lot of work to make it right :)

File: packages/core-data/src/entities.js (hunk 3 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js]: Done in https://github.com/WordPress/gutenberg/pull/72145/commits/f332ffb7601e4a9b8d0f8020c52843d07a0f9645

File: packages/core-data/src/entities.js (hunk 2 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js]: Implemented in https://github.com/WordPress/gutenberg/pull/72145/commits/3cef376a6056aa90d9c53fd98e61849d9ca44fe8, with a test in https://github.com/WordPress/gutenberg/pull/72145/commits/805e9b22bed43e1950f888ad3eef8a0eb240cd47.

I'll confirm @ingeniumed is credited with props for the initial resolver tests.

**Review by youknowriad — APPROVED** [review]
This is cool, I like where this is going.

**Comment by chriszarate** [comment]
@youknowriad The [Core Backport action](https://github.com/WordPress/gutenberg/actions/runs/18351160590/job/52271227808?pr=72145) is failing. I believe this is due to [this change to a PHPUnit test case](https://github.com/WordPress/gutenberg/pull/72145/commits/1804c116e25368ecbaff757eb3d3ab1f7b7537e3).

However, that unit test does not exist in WP Core. Is it ok to add the `No Core Sync Required` label [per the docs](https://github.com/WordPress/gutenberg/blob/033b64ed0594e9af6c209385c5b8f6a8dfbc81b4/backport-changelog/readme.md?plain=1#L57)?

**Comment by chriszarate** [comment]
> I believe this is due to [this change to a PHPUnit test case](https://github.com/WordPress/gutenberg/pull/72145/commits/1804c116e25368ecbaff757eb3d3ab1f7b7537e3).

I'm not even sure why that change is required for unit tests to pass, but it is. I am guessing it has something to do with changing the `sync` package's entrypoint to TypeScript, but I'm not sure.

**Comment by chriszarate** [comment]
@youknowriad Nevermind, I realized that this PHP unit test requires a full `npm run build` to reflect local changes. Ignore with apologies.

**Comment by youknowriad** [comment]
So I think this PR changed a bit more things that necessary. It seems to have moved the "block parsing" from "editor loading" to "entity fetching", which means any block editor that tries to lazy load the block assets / block registration is broken (returns an empty block lists for any post because no block is registered). This happens in the extensible site editor for instance.

I wonder if that change also has a performance penalty. How possible would it be to make block reading/parsing more "lazy"?

**Comment by chriszarate** [comment]
> It seems to have moved the "block parsing" from "editor loading" to "entity fetching"

I assume you refer to the block parsing (via transient config `read`) that happens in `getEntityRecord` (corresponding to "entity fetching"). What `core-data` action corresponds to "editor loading"? Or do we need to invent one?

**Comment by youknowriad** [comment]
>  What core-data action corresponds to "editor loading"? Or do we need to invent one?

We don't have one, we used to do that in `useEntityBlockEditor`
